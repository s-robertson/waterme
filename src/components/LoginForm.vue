<template>
  <div class="login-form">
    <ErrorList :errors="loginErrors" />
    <form action="" @submit.prevent="handleSubmit">
      <div class="c-form-item">
        <label
          >Email:
          <input v-model="email" type="text" data-test="input-login-email"
        /></label>
      </div>
      <div class="c-form-item">
        <label
          >Password:
          <input v-model="password" type="text" data-test="input-login-password"
        /></label>
      </div>
      <OptionButtons>
        <button type="submit" data-test="input-login-submit">Login</button>
        <router-link to="/register" class="button"
          >Register for a New Account</router-link
        >
      </OptionButtons>
    </form>
  </div>
</template>

<script>
import { mapActions, mapState } from "vuex";
import ErrorList from "@/components/ErrorList";
import OptionButtons from "@/components/OptionButtons";

export default {
  name: "LoginForm",
  components: { OptionButtons, ErrorList },
  data() {
    return {
      email: "",
      password: ""
    };
  },
  computed: {
    ...mapState(["loginErrors"])
  },
  methods: {
    ...mapActions(["login"]),
    async handleSubmit() {
      const loginResult = await this.login({
        email: this.email,
        password: this.password
      });

      if (loginResult) {
        this.$router.push("/");
      }
    }
  }
};
</script>
