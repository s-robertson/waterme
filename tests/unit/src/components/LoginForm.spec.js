import { createLocalVue, shallowMount } from "@vue/test-utils";
import LoginForm from "@/components/LoginForm";
import Vuex from "vuex";
import { createStore } from "../../helpers";

const localVue = createLocalVue();

localVue.use(Vuex);

describe("LoginForm component", () => {
  const defaultStoreOptions = {
    state: {
      loginErrors: []
    },
    actions: {
      login: jest.fn()
    }
  };

  it("Should render correctly", () => {
    const wrapper = shallowMount(LoginForm, {
      localVue,
      store: createStore(defaultStoreOptions),
      stubs: ["router-link"]
    });

    expect(wrapper.html()).toMatchSnapshot();
  });

  it("Should render any login errors", () => {
    const wrapper = shallowMount(LoginForm, {
      localVue,
      store: createStore(defaultStoreOptions, {
        state: {
          loginErrors: ["Invalid password"]
        }
      }),
      stubs: ["router-link"]
    });

    expect(wrapper.html()).toMatchSnapshot();
  });

  it("Should log user in", async () => {
    const expectedEmail = "test@user.com";
    const expectedPassword = "myPassword123";
    const $router = {
      push: jest.fn()
    };
    defaultStoreOptions.actions.login.mockImplementation(async () => true);

    const wrapper = shallowMount(LoginForm, {
      localVue,
      store: createStore(defaultStoreOptions),
      mocks: {
        $router
      },
      stubs: ["router-link"]
    });

    const inputs = wrapper.findAll("input[type=text]");

    inputs.at(0).setValue(expectedEmail);
    inputs.at(1).setValue(expectedPassword);

    wrapper.find("form").trigger("submit");
    await wrapper.vm.$nextTick();

    expect(defaultStoreOptions.actions.login).toHaveBeenCalledWith(
      expect.anything(),
      {
        email: expectedEmail,
        password: expectedPassword
      }
    );
    expect($router.push).toHaveBeenCalledWith("/");
  });
});
