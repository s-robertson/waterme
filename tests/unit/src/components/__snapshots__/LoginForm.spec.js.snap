// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LoginForm component Should render any login errors 1`] = `
<div class="login-form">
  <errorlist-stub errors="Invalid password"></errorlist-stub>
  <form action="">
    <div class="c-form-item"><label>Email:
        <input type="text" data-test="input-login-email"></label></div>
    <div class="c-form-item"><label>Password:
        <input type="text" data-test="input-login-password"></label></div>
    <optionbuttons-stub><button type="submit" data-test="input-login-submit">Login</button>
      <router-link-stub to="/register" class="button">Register for a New Account</router-link-stub>
    </optionbuttons-stub>
  </form>
</div>
`;

exports[`LoginForm component Should render correctly 1`] = `
<div class="login-form">
  <errorlist-stub errors=""></errorlist-stub>
  <form action="">
    <div class="c-form-item"><label>Email:
        <input type="text" data-test="input-login-email"></label></div>
    <div class="c-form-item"><label>Password:
        <input type="text" data-test="input-login-password"></label></div>
    <optionbuttons-stub><button type="submit" data-test="input-login-submit">Login</button>
      <router-link-stub to="/register" class="button">Register for a New Account</router-link-stub>
    </optionbuttons-stub>
  </form>
</div>
`;
